### PR TITLE
Don't blindly assume input file has .aes extension

### LINKED ISSUE #9: Segmentation fault
I think it would be better if you don't assume that the file has the ".aes" extension, because with a file with a name of 4 or less characters, the program crashes.

### PR DESCRIPTION
**Pull request by skeeto**
Fixes #9. Ran into this myself. I was intending to only open an issue for it, but seeing that you knew about it and that it still wasn't fixed, I went ahead and fixed it.

### CODE REVIEW

File: Linux/src/aescrypt.c [hunk 1]
```diff
@@ -1192,9 +1193,17 @@ int main(int argc, char *argv[])
         {
             if (outfp == NULL)
             {
-                /* assume .aes extension */
-                strncpy(outfile, infile, strlen(infile)-4);
-                outfile[strlen(infile)-4] = '\0';
+                /* input file name must have .aes extension */
+                infile_len = strlen(infile);
+                if (infile_len < 5 || strcmp(infile + infile_len - 4, ".aes"))
```
Comment (paulej) [inline Linux/src/aescrypt.c]: This line produces a compiler warning for me:
```
aescrypt.c: In function ‘main’:
aescrypt.c:1205:17: warning: ‘strncpy’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
                 strncpy(outfile, infile, infile_len-4);
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
aescrypt.c:1197:30: note: length computed here
                 infile_len = strlen(infile);
                              ^~~~~~~~~~~~~~
mv -f .deps/aescrypt.Tpo .deps/aescrypt.Po
```

A solution is to move the strlen() inside the function and get rid of hte infile_len variable.  I think the code is right, but we need to change it so the compiler does not complain.

**Comment by skeeto** [comment]
This is a problem with the original code. I preserved it in my change 
here since it wasn't the specific problem I was trying to fix. If you're 
seeing the warning now but not before, it's probably because the issue 
wasn't visible to the compiler until I replaced the strlen() with a 
variable.

The problem is that outfile is a fixed length, and the third parameter 
to strncpy() is incorrect. It should be "sizeof(outfile)" and the manual 
null-termination fixed. I left it alone since, ultimately, this is just 
trading a bad bug for a less bad bug. Instead of overflowing, the 
filename is silently truncated to the wrong filename.

I just pushed another commit to fixup all the usages of outfile, but 
leaving the silent truncation.

**Comment by paulej** [comment]
gcc is just being picky.  It doesn't like `strcmp(infile + infile_len - 4, ".aes")` but is perfectly OK with `strcmp(infile + strlen(infile) - 4, ".aes")`.   In this instance, they are identical, but one causes a warning and the other does not.  If you're not seeing the warning, then perhaps I'll do a manual merge and do whatever I need to make gcc stop whining. :)

**Comment by skeeto** [comment]
Ah, I see. Yeah, the strcmp() one is a false positive, and I'm unable to 
reproduce it. Since that's a GCC bug, it might be limited to older 
versions of GCC. The warning printout you showed before is complaining 
about the third argument to strncpy(), which _is_ a legitimate issue — 
though, again, for some reason I couldn't reproduce that either (GCC 
8.3.0, built myself from source). Clang doesn't have -Wstringop-overflow.

**Pull request closed without merging**